tools/pythonpkg/src/vector_conversion.cpp

```diff
@@ -411,6 +413,68 @@ void VectorConversion::NumpyToDuckDB(PandasColumnBindData &bind_data, py::array
 	}
 }
 
+void VectorConversion::BindNumpy(const DBConfig &config, py::handle df, vector<PandasColumnBindData> &bind_columns,
+                                 vector<LogicalType> &return_types, vector<string> &names) {
+	auto df_columns = py::list(df.attr("keys")());
+	auto df_types = py::list();
+	for (auto item : py::cast<py::dict>(df)) {
+		if (string(py::str(item.second.attr("dtype").attr("char"))) == "U") {
+			df_types.attr("append")(py::str("string"));
+			continue;
+		}
+		df_types.attr("append")(py::str(item.second.attr("dtype")));
+	}
+	auto get_fun = df.attr("__getitem__");
+	if (py::len(df_columns) == 0 || py::len(df_types) == 0 || py::len(df_columns) != py::len(df_types)) {
+		throw InvalidInputException("Need a DataFrame with at least one column");
+	}
+	for (idx_t col_idx = 0; col_idx < py::len(df_columns); col_idx++) {
+		LogicalType duckdb_col_type;
+		PandasColumnBindData bind_data;
+
+		names.emplace_back(py::str(df_columns[col_idx]));
+		bind_data.pandas_type = ConvertPandasType(df_types[col_idx]);
+
+		auto column = get_fun(df_columns[col_idx]);
+
+		if (bind_data.pandas_type == PandasType::FLOAT_16) {
+			bind_data.numpy_col = py::array(column.attr("astype")("float32"));
+			bind_data.pandas_type = PandasType::FLOAT_32;
+			duckdb_col_type = PandasToLogicalType(bind_data.pandas_type);
+		} else if (bind_data.pandas_type == PandasType::OBJECT && string(py::str(df_types[col_idx])) == "string") {
+			bind_data.pandas_type = PandasType::CATEGORY;
+			auto enum_name = string(py::str(df_columns[col_idx]));
+			auto uniq = py::cast<py::tuple>(py::module_::import("numpy").attr("unique")(column, false, true));
```

Comment:
I feel like this needs a comment to explain that this produces a tuple containing the distinct entries (0) and the indices into that array for the corresponding original values (1)